### ISSUE TITLE
NDK issue after flutter upgrade, No version of NDK matched the requested version 20.0.5594570. Versions available locally: 21.0.6113669

### ISSUE THREAD

**Issue by r4v3r23** [flutter/flutter#53657]
I upgraded flutter master channel yesterday and now I receiving this message when installing:

`flutter run`

> FAILURE: Build failed with an exception.                                
>                                                                         
> * What went wrong:                                                      
> Execution failed for task ':app:stripDebugDebugSymbols'.                
> 
> No version of NDK matched the requested version 20.0.5594570. Versions available locally: 21.0.6113669

how do I fix this?

**Comment by VladyslavBondarenko** [comment]
@r4v3r23 
Could you provide your `flutter doctor -v`?

Related to https://github.com/flutter/flutter/issues/27967, https://github.com/gradle/gradle/issues/12440

**Comment by r4v3r23** [comment]
> @r4v3r23
> Could you provide your `flutter doctor -v`?
> 
> Related to #27967, [gradle/gradle#12440](https://github.com/gradle/gradle/issues/12440)

✓] Flutter (Channel master, v1.16.4-pre.22, on Linux, locale en_US.UTF-8)
    • Flutter version 1.16.4-pre.22 at /home/x0x/Android/flutter
    • Framework revision e4a61dfdea (2 days ago), 2020-03-29 21:01:02 -0400
    • Engine revision 361d5d5320
    • Dart version 2.8.0 (build 2.8.0-dev.18.0 1402e8e1a4)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /home/x0x/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Android Studio (version 3.6)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.

**Comment by jmagman** [comment]
> Android NDK location not configured (optional; useful for native profiling support)

https://stackoverflow.com/a/60413117 and https://github.com/gradle/gradle/issues/12440#issuecomment-606188282

@r4v3r23 What's the output of `flutter doctor -vv` (note the extra v)?

**Comment by r4v3r23** [comment]
> > Android NDK location not configured (optional; useful for native profiling support)
> 
> https://stackoverflow.com/a/60413117 and [gradle/gradle#12440 (comment)](https://github.com/gradle/gradle/issues/12440#issuecomment-606188282)
> 
> @r4v3r23 What's the output of `flutter doctor -vv` (note the extra v)?

flutter doctor -vv
[  +14 ms] executing: [/home/USER/Android/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +21 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] e4a61dfdea8dc55e6665abefb00be9575de4d06c
[        ] executing: [/home/USER/Android/flutter/] git describe --match v*.*.* --first-parent --long --tags
[  +12 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.16.3-22-ge4a61dfde
[   +6 ms] executing: [/home/USER/Android/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/master
[        ] executing: [/home/USER/Android/flutter/] git ls-remote --get-url origin
[   +2 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +50 ms] executing: [/home/USER/Android/flutter/] git rev-parse --abbrev-ref HEAD
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] master
[  +22 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +2 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[   +6 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
[        ] Artifact Instance of 'GradleWrapper' is not required, skipping update.
[        ] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[   +2 ms] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ] Artifact Instance of 'FlutterSdk' is not required, skipping update.
[        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping update.
[  +18 ms] executing: [/home/USER/Android/flutter/] git rev-parse --abbrev-ref HEAD
[   +4 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] master
[        ] executing: [/home/USER/Android/flutter/] git fetch https://github.com/flutter/flutter.git --tags
[+15450 ms] Exit code 0 from: git fetch https://github.com/flutter/flutter.git --tags
[   +1 ms] From https://github.com/flutter/flutter
            * branch                HEAD           -> FETCH_HEAD
            * [new tag]             1.17.0-dev.0.0 -> 1.17.0-dev.0.0
[        ] executing: [/home/USER/Android/flutter/] git describe --match v*.*.* --first-parent --long --tags
[  +15 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.16.3-22-ge4a61dfde
[   +4 ms] Doctor summary (to see all details, run flutter doctor -v):
[  +11 ms] executing: /opt/android-studio/jre/bin/java -version
[  +55 ms] Exit code 0 from: /opt/android-studio/jre/bin/java -version
[        ] openjdk version "1.8.0_212-release"
           OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
           OpenJDK 64-Bit Server VM (build 25.212-b4-5784211, mixed mode)
[  +12 ms] executing: /opt/android-studio/jre/bin/java -version
[  +54 ms] Exit code 0 from: /opt/android-studio/jre/bin/java -version
[        ] openjdk version "1.8.0_212-release"
           OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
           OpenJDK 64-Bit Server VM (build 25.212-b4-5784211, mixed mode)
[   +7 ms] executing: [/home/USER/Android/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%ar
[   +3 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%ar
[        ] 2 days ago
[  +12 ms] executing: /home/SER/Android/flutter/bin/cache/artifacts/engine/android-arm-profile/linux-x64/gen_snapshot
[   +3 ms] Exit code 255 from: /home/USER/Android/flutter/bin/cache/artifacts/engine/android-arm-profile/linux-x64/gen_snapshot
[        ] At least one input is required
           Usage: gen_snapshot [<vm-flags>] [<options>] <dart-kernel-file>             

           Common options:                                                             
           --help                                                                      
             Display this message (add --verbose for information about all VM options).
           --version                                                                   
             Print the VM version.                                                     

           To create a core snapshot:                                                  
           --snapshot_kind=core                                                        
           --vm_snapshot_data=<output-file>                                            
           --isolate_snapshot_data=<output-file>                                       
           <dart-kernel-file>                                                          

           To create an AOT application snapshot as assembly suitable for compilation  
           as a static or dynamic library:                                             
           --snapshot_kind=app-aot-assembly                                            
           --assembly=<output-file>                                                    
           [--strip]                                                                   
           [--obfuscate]                                                               
           [--save-debugging-info=<debug-filename>]                                    
           [--save-obfuscation-map=<map-filename>]                                     
           <dart-kernel-file>                                                          

           To create an AOT application snapshot as an ELF shared library:             
           --snapshot_kind=app-aot-elf                                                 
           --elf=<output-file>                                                         
           [--strip]                                                                   
           [--obfuscate]                                                               
           [--save-debugging-info=<debug-filename>]                                    
           [--save-obfuscation-map=<map-filename>]                                     
           <dart-kernel-file>                                                          

           AOT snapshots can be obfuscated: that is all identifiers will be renamed    
           during compilation. This mode is enabled with --obfuscate flag. Mapping     
           between original and obfuscated names can be serialized as a JSON array     
           using --save-obfuscation-map=<filename> option. See dartbug.com/30524       
           for implementation details and limitations of the obfuscation pass.
[  +11 ms] java -version
[ +132 ms] executing: /home/USER/Android/Sdk/platform-tools/adb devices -l
[  +19 ms] [✓] Flutter (Channel master, v1.16.4-pre.22, on Linux, locale en_US.UTF-8)
⣽[  +38 ms] List of devices attached
[   +6 ms] executing: /home/USER/Android/Sdk/platform-tools/adb devices -l
[   +7 ms] List of devices attached
[  +28 ms] executing: /home/USER/Android/Sdk/cmdline-tools/latest/bin/sdkmanager --licenses
 
[+2403 ms] [✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[   +2 ms] [✓] Android Studio (version 3.6)
[   +1 ms] [!] Connected device
[        ]     ! No devices available
[        ] ! Doctor found issues in 1 category.
[  +12 ms] "flutter doctor" took 18,357ms.

**Comment by r4v3r23** [comment]
@jmagman I set NDK location in project structure but still get same result

**Comment by jmagman** [comment]
That output didn't have what I was hoping for, filed https://github.com/flutter/flutter/issues/53680.

**Comment by canewsin** [comment]
I was seeing this behaviour on latest stable 
```
[√] Flutter (Channel stable, v1.12.13+hotfix.8, on Microsoft Windows [Version 10.0.18363.720], locale en-IN)
    • Flutter version 1.12.13+hotfix.8 at C:\flutter\1.0.0
    • Framework revision 0b8abb4724 (7 weeks ago), 2020-02-11 11:44:36 -0800
    • Engine revision e1e6ced81d
    • Dart version 2.7.0


[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at D:\Android\SDK
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = D:\Android\SDK
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
    • All Android licenses accepted.

[√] Android Studio (version 3.6)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)

[√] VS Code (version 1.43.2)
    • VS Code at C:\Users\PramUkesh\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.8.1

[√] Connected device (1 available)
    • Redmi K20 Pro • 270a97e6 • android-arm64 • Android 10 (API 29)

• No issues found!
```

**Comment by r4v3r23** [comment]
> That output didn't have what I was hoping for, filed #53680.

so whats the move here?

**Comment by jmagman** [comment]
The NDK doctor output was unrelated and unused, so you can ignore that.

Flutter doesn't configure the app:stripDebugDebugSymbols Gradle task.  Based on https://github.com/gradle/gradle/issues/12440#issuecomment-606188282 it looks like a Gradle bug. 
 It's possible you have some mismatch between your SDK and plugin.

@blasten do you have any ideas?

**Comment by r4v3r23** [comment]
> The NDK doctor output was unrelated and unused, so you can ignore that.
> 
> Flutter doesn't configure the app:stripDebugDebugSymbols Gradle task. Based on [gradle/gradle#12440 (comment)](https://github.com/gradle/gradle/issues/12440#issuecomment-606188282) it looks like a Gradle bug. Did any of the workarounds in that comment help?

deleting NDK folder worked.

I am still only able to build --debug versions of the app
'flutter build apk --release' throws loads of errors

**Comment by jmagman** [comment]
> I am still only able to build --debug versions of the app
> 'flutter build apk --release' throws loads of errors

Can you file a new issue for that?  Let's keep this one to see if other people are hitting the NDK bug.  This is the only report I'm aware of.

**Comment by ThinkDigitalSoftware** [comment]
Deleting the NDK folder worked for me too. I had also previously accepted the ndk upgrade suggestion in Android Studio, something I don't usually do since I don't open Android Studio often

**Comment by deacon78** [comment]
same issue after upgrading graddle :(

**Comment by UsmaJulian** [comment]
Only need go to tools-sdkmanager-sdktools-uncheck hide obsolete packages-check show package details-select ndk version and apply ____ done.

**Comment by marchacio** [comment]
I solved it by bringing the version of the grandle to 3.5.0

**In android/build.grandle**
```
buildscript {
    repositories {
        google()
        jcenter()
    }

    dependencies {
        classpath 'com.android.tools.build:gradle:3.5.0' //Changed this line
    }
}
```
Hope i help you guys!


> I upgraded flutter master channel yesterday and now I receiving this message when installing:
> 
> `flutter run`
> 
> > FAILURE: Build failed with an exception.
> > 
> > * What went wrong:
> >   Execution failed for task ':app:stripDebugDebugSymbols'.
> > 
> > No version of NDK matched the requested version 20.0.5594570. Versions available locally: 21.0.6113669
> 
> how do I fix this?



> same issue after upgrading graddle :(

**Comment by larryaasen** [comment]
> I solved it by bringing the version of the grandle to 3.5.0
> 

Using gradle 3.5.0 worked for me too. It had been set to 3.6.3 which did not work.

**Comment by mahmoudalaa97** [comment]
> I solved it by bringing the version of the grandle to 3.5.0
> 
> **In android/build.grandle**
> 
> ```
> buildscript {
>     repositories {
>         google()
>         jcenter()
>     }
> 
>     dependencies {
>         classpath 'com.android.tools.build:gradle:3.5.0' //Changed this line
>     }
> }
> ```
> 
> Hope i help you guys!
> 
> > I upgraded flutter master channel yesterday and now I receiving this message when installing:
> > `flutter run`
> > > FAILURE: Build failed with an exception.
> > > 
> > > * What went wrong:
> > >   Execution failed for task ':app:stripDebugDebugSymbols'.
> > > 
> > > No version of NDK matched the requested version 20.0.5594570. Versions available locally: 21.0.6113669
> > 
> > 
> > how do I fix this?
> 
> > same issue after upgrading graddle :(

worked to me when i changed to this version thanks bro :) 👍

**Comment by duladissa** [comment]
You need to install the exact version(**20.0.5594570**) requested in error from **Android Studio-> Preferences -> Android SDK -> SDK Tools -> NDK(Side by side)**
(To view all the NDK versions, tick "Show Package Details")

**No version of NDK matched the requested version 20.0.5594570. Versions available locally: 20.1.5948944, 21.2.6472646**

<img width="1030" alt="Screenshot 2020-05-28 at 9 45 12 AM" src="https://user-images.githubusercontent.com/7032434/83089207-5d73e080-a0c8-11ea-8f70-fd36394ebb3e.png">

**Comment by wendreof** [comment]
I fixed the error, opening the android folder on Android Studio, **rebuild** -> **Download new NDK version**.

**Comment by nawalhmw** [comment]
I was facing the same issue & was able to fix it by downloading the NDK & manually downgrading my project's Gradle version in the **android/build.gradle** from 4.0.1 to 3.5.0.

`dependencies {`
`        classpath 'com.android.tools.build:gradle:3.5.0' //edited this`
`    }`

**Comment by deakjahn** [comment]
As mentioned above, there is no need to downgrade. Just specifiy the correct NDK to use.

**Comment by shorben07** [comment]
Downgrade to Android Gradle plugin 3.5.0 works because it is the last version when the default NDK version is not specified. 
<img width="904" alt="Screenshot 2020-11-13 at 10 00 49" src="https://user-images.githubusercontent.com/36268266/99043656-35edd300-2597-11eb-9e70-b063d305762f.png">
https://developer.android.com/studio/projects/install-ndk#default-ndk-per-agp

**Comment by deakjahn** [comment]
I really can't understand why you would want to downgrade and potentially lose functionality (or bugfixes or whatever) when it's really as simple as a single line in your build file. But OK, each to their own. :-)

**Comment by predatorx7** [comment]
The same issue is happening with me too. I can't downgrade android gradle to 3.5.0 due to an important dependency.

**Comment by deakjahn** [comment]
@predatorx7 Just specify the current NDK version in `app/build.gradle`.

**Comment by smkhalsa** [comment]
This is a real pain in the neck when using CI / CD. Any suggestions besides downgrading gradle or manually setting the   ndkVersion?

**Comment by xster** [comment]
Thanks all for your patience. I hadn't realized the amount of GitHub reactions there were in the response posts of the thread. I did a bit of digging and here's what I think is going on. I might not be 100% right but I think I'm close.

Fundamentally this is a "bug" (or at least an ergonomic bug) in the Android Gradle Plugin that all Android projects use and theoretically should affect all Android projects that use native libraries. Flutter doesn't directly use the NDK but if any NDK is installed, it then uses an NDK to map debug symbols for stripped native libraries when debugging. The usability regression happened between AGP versions 3.6.0-beta02 (and thus includes release versions of 3.6.0) all the way through to 4.1.0. In those versions, as @shorben07 [mentioned above](https://github.com/flutter/flutter/issues/53657#issuecomment-726590816), each specific AGP version prefers a specific associated NDK version. That's fine but it's a hassle if that specific NDK version isn't installed locally and AGP doesn't install it for you. AGP 4.1.0 fixes that bug in https://issuetracker.google.com/issues/144111441. 

Up until Flutter 2.0.0, the default template's AGP version was 3.5.0 so most Flutter users weren't running into this issue. You're likely running into this issue if you opened the <project root>/android folder in Android Studio before April 2020 (when AGP 4.0.0 was released), and clicked `update` in the auto AGP upgrade prompt pop-up. 

![Screen Shot 2021-03-10 at 9 05 17 AM](https://user-images.githubusercontent.com/156888/110670177-f3718700-8181-11eb-8431-839841a9edfa.png)

Or if you did the manual equivalent by editing the <project root>/android/build.gradle file and changing the version in:

```groovy
dependencies {
    classpath 'com.android.tools.build:gradle:[a version >=3.6.0]'
}
```

If you auto-updated AGP after April 2020 and before Flutter 2.0.0, you would have run into https://github.com/flutter/flutter/issues/58247 anyway since Flutter didn't support AGP version 4+ until Flutter 2.0.0. 

For all new projects created after Flutter 2.0.0, this issue doesn't occur either since new projects use AGP 4.1.0 since https://github.com/flutter/flutter/pull/71446. 

If you develop using add-to-app, you could also run into this issue if your host project that brings in a Flutter module uses an AGP version between 3.6 and 4.0 inclusively. 

If you're having issues currently, you can:

1- Go to your SDK manager, SDK Tools tab (click Show Package Details if you want to see which versions you have) and generally just remove the entire NDK category. AGP shouldn't make any attempts to desymbolicate native libraries if there aren't any NDKs installed
![Screen Shot 2021-03-10 at 9 05 45 AM](https://user-images.githubusercontent.com/156888/110670581-6975ee00-8182-11eb-8fd0-c77ae98cd6bf.png)
This method saves you disk space but may not be desirable if you have legitimate NDK usages for your other Android projects. 
2- **[Recommended]** Go to your <project root>/android/build.gradle and change your AGP version to 
```groovy
dependencies {
    classpath 'com.android.tools.build:gradle:4.1.2'
}
```
which is the latest stable release of AGP with https://issuetracker.google.com/issues/144111441 fixed. Then you shouldn't be manually bothered to install specific NDK versions anymore. 
3- If you need to stay with your current AGP version for some reason, then manually install the requested NDK version in the SDK manager. For instance, if the error says `No version of NDK matched the requested version 20.0.5594570. Versions available locally: 21.0.6113669`, then install version 20.0.5594570. Don't forget to tick "Show Package Details" in the dialog. This method isn't ideal since each time you take your code onto a new machine, you have to make sure that machine has the needed NDK version installed. This is especially problematic for CI where you don't know what NDK version is installed in the cloud unless you built the docker image yourself. But if you're developing locally, that NDK version should be stable as long as you don't change your AGP version. 
4- **[Not recommended]** You can pin a specific requested NDK version as suggested above https://github.com/flutter/flutter/issues/53657#issuecomment-642236438. It's ok if you just develop with Flutter (but if that's the case, you have no other restrictions and might as well use approach 1 or 2) but if you really use the NDK, each AGP version is specifically tested for compatibility with one NDK version, and you're losing that testing if you "mismatch" the version. Plus you get all the disadvantages of 3.

**Comment by Zakria11223** [comment]
> > I solved it by bringing the version of the grandle to 3.5.0
> > **In android/build.grandle**
> > ```
> > buildscript {
> >     repositories {
> >         google()
> >         jcenter()
> >     }
> > 
> >     dependencies {
> >         classpath 'com.android.tools.build:gradle:3.5.0' //Changed this line
> >     }
> > }
> > ```
> > 
> > 
> > Hope i help you guys!
> > > I upgraded flutter master channel yesterday and now I receiving this message when installing:
> > > `flutter run`
> > > > FAILURE: Build failed with an exception.
> > > > 
> > > > * What went wrong:
> > > >   Execution failed for task ':app:stripDebugDebugSymbols'.
> > > > 
> > > > No version of NDK matched the requested version 20.0.5594570. Versions available locally: 21.0.6113669
> > > 
> > > 
> > > how do I fix this?
> > 
> > 
> > > same issue after upgrading graddle :(
> 
> worked to me when i changed to this version thanks bro :) 👍
May Allah Bless you. After doing this, my project is running.

**Comment by chhinsras** [comment]
> @duladissa, no, you don't have to. No need to downgrade. Just put the NDK you wish to use into your `build.gradle` (the one inside `app`, not the one above):
> 
> ```
> android {
>   compileSdkVersion 29
>   ndkVersion '21.3.6528147'
>   ...
> }
> ```
> 
> Still, it's a pain to change every time NDK upgrades. Flutter team, it's time to do something about it. :-)

Perfect! After spending 4 hours
